Rook StorageClassDeviceSet design doc

jarrpa opened the pull request:
A design doc for #3305 to define the notion of DeviceSets in Rook, with an example for rook-ceph.

Related: #919 

Signed-off-by: Jose A. Rivera <jarrpa@redhat.com>

[skip ci]

leseb commented on design/ceph-device-set.md:
```diff
@@ -0,0 +1,86 @@
+
+If the admin wishes to control device placement beyond simply using node hostnames it will be up to them to make sure the desired nodes are labeled properly to ensure the Kuberenetes scheduler will distribute the OSD Pods appropriately.
+
+Additionally, it is recommended that the admin setup a StorageClass with `volumeBindingMode: WaitForFirstConsumer` set, and that they use this StorageClass for providing the OSD PVCs.
```
```suggestion
Additionally, it is recommended that the admin setups a StorageClass with `volumeBindingMode: WaitForFirstConsumer` set, and that they use this StorageClass for providing the OSD PVCs.
```

jarrpa replied:
This change doesn't make grammatical sense... unless you mean to replace it with `sets up`, but that's not quite grammatically correct in English either.

leseb replied:
Ok never mind then.

jarrpa commented:
@jbw976 I was actually thinking about that, but I don't know much about Rook outside of rook-ceph to know how to best go about that. Do you want to give some suggestions? Would it be sufficient just to take "Ceph" out of the name and change the API it belongs to?

@leseb I see mons being described in the CephCluster CRD itself. They would probably also follow a similar StatefulSet model, but not be defined in a single block like this CRD. I imagine they would still inherit from the cluster-level configs in `spec.placement.mons` and `spec.resources.mons` along with a new field called something like `spec.mons.volumeClaimTemplate` (which is only a single template, not a list). I believe something like that belongs in it's own PR, even if ultimately they share code underneath. But if you think it makes sense here, I can add another document to this PR to describe that.

kshlm commented:
If all the CephDeviceSet controller does is create and manage StatefulSets of OSDs, do we really need to have it as a new controller? To me, this would be better handled by the OSD controller. It already launches and manages OSD pods based on the storage information provided. DeviceSets are just another way to provide this information, using which the OSD controller can now launch StatefulSets instead of individual pods.

Having a separate DeviceSet controller would be useful if we wanted to DeviceSets with other operators. The DeviceSet CRD could be extended to take a Pod template parameter, and the controller would launch StatefulSets with the provided template. Other storage operators/controllers could then create a DeviceSet CR with their own pod templates to launch their storage pods. For eg. this could allow the ceph-mon operator to use DeviceSets.

But then again, the storage controllers could just create the StatefulSets by themselves instead of creating the DeviceSet CR.

The more I look at DeviceSets, it seems like a just a spec to me. The spec defines a set of storage devices available to the storage controllers. It is upto the storage controllers to make use of the storage devices as they see fit.

rohan47 commented on design/ceph-device-set.md:
````diff
@@ -0,0 +1,86 @@
+          topologyKey: "failure-domain.beta.kubernetes.io/zone"
+```
+
+The CephDeviceSet CRD will define characteristics of a set of OSD devices and their Pods. The following fields are available in the spec:
````
Instead of `CephDeviceSet` CRD is it possible to have `DeviceSet` CRD which will define characterstics of a set of devices which can be consumed by the storage providers. For example a DeviceSet has a count 3, Ceph can use it for backing 3 OSDs, NFS operator can use it for creating 3 exports, cassandra can use it for its 3 rack members(cassandra nodes)

rohan47 replied:
Also as per @travisn comment https://github.com/rook/rook/pull/3306#discussion_r294321699 if the pattern for OSD is not changed, this makes a good general functionality.

jarrpa commented:
@jbw976 @kshlm @rohan47 I updated the design to make the following major changes:

* Rename CephDeviceSet to DeviceSet
* Propose DeviceSet as a struct in Rook

@leseb I still omitted a mention of mons, since I believe those are better served not using the DeviceSet struct directly. They may still share the same implementation of using PVCs that OSDs will, but that's beyond the scope of this design.

@travisn I still kept the OSDs managed by StatefulSets because I'm convinced that it is the simplest, cleanest way to manage the Pods and PVCs using as much of the Kubernetes infrastructure as possible. I understand that this does involve drastically departing from the way things are done now, but I believe the ease in implementation will be worth it.

If you can write down a workflow using individual Deployments per OSD that you think is comparably reasonable I'm all for it, but I can't imagine such at this time. The way I'm seeing it, we would have to do something like:
1. Get all matching OSD PVCs
1. Create any missing OSD PVCs
1. Get all matching OSD Deployments/Pods
1. Check that all OSD Deployments are using valid OSD PVCs
    * If not, probably remove the OSD Deployment? Or just skip with an error?
1. Take the list of PVCs without Deployments, check if an OsdPrepare Job is using it
1. Run OsdPrepare on all unused and uninitialized PVCs
    * This may be either one Job per PVC or one Job per batch of PVCs. One per PVC sounds cleaner.
1. Create an OSD Deployment for each unused but initialized PVC

Am I not seeing something that would significantly compact all that?

BlaineEXE commented on design/device-set.md:
````diff
@@ -0,0 +1,50 @@
+
+## DeviceSet struct
+```go
+struct DeviceSet {
````
It seems to me that the purpose of `DeviceSet`s is to allow Rook to use storage classes as backing storage. I don't see any use for this beyond consumption of storage classes for devices. To this end, I think DeviceSet isn't the best name for this. It isn't very specific, and it seems like it would be a lot clearer to users if we name this something like `StorageClassDeviceSet`.

jarrpa replied:
Not a bad idea.

jarrpa commented:
Updated the design again to rename the struct per @BlaineEXE's suggestion to `StorageClassDeviceSet`. Additionally added a generic `Config` field, which would be just a `map[string]string` to allow for provider-specific configuration.

jarrpa commented:
During today's Rook community meeting, it was announced that we would be holding another meeting tomorrow to dive into deep discussion on this design. It will be held over BlueJeans, so do try to test it out ahead of time if you haven't used it before.

**Subject:** Rook StorageClassDeviceSet design discussion
**Time:** Wednesday June 19th @ 8:30 PST / 9:30 MST / 10:30 CST / 11:30 EST / 17:30 CEST / 21:00 IST
**Location:** https://redhat.bluejeans.com/182997528

noahdesu commented on design/ceph-device-set.md:
```diff
@@ -0,0 +1,86 @@
+Additionally, it is recommended that the admin setup a StorageClass with `volumeBindingMode: WaitForFirstConsumer` set, and that they use this StorageClass for providing the OSD PVCs.
+
+### OSDs as StatefulSets
+In this design, a set of OSDs will be represented by a StatefulSet with the following configuration:
```
IIUC the StatefulSet in principle would make things easier because it maintains a strict affinity between a PV, pod, and replica index. Presumably a ConfigMap would be able to translate between a replica index and an OSD ID freeing Rook from correctly managing the affinity between this state as well as some aspects of scheduling?

I think it would be valuable here to enumerate the useful properties of StatefulSets independent of the workload object type ultimately chosen.

jarrpa replied:
The benefits of a StatefulSet include not only the thing with the ConfigMap you just described, but all the work I outlined in https://github.com/rook/rook/pull/3306#issuecomment-502916870 with a single, well-known Kubernetes primitive. I don't know how to further elaborate on the benefits, so I'll just link to the Kubernetes page on them: https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/

noahdesu replied:
Ah, didn't read the last part of that comment. That's helpful thanks

noahdesu commented:
Is PV an abstraction that can encompass the current approach of using bind mounts to map raw devices into the OSD container? Or... will this be a separate dimension to scheduling and configuration that must be maintained?

jarrpa commented:
@noahdesu I forget the exact implementation, all I know is that a PV can be configured with `volumeMode: Block`. When this is done, Kubernetes will attach the underlying storage to the node the Pod is scheduled on and present it as a raw block device inside the container at a path you specify. I think it uses bind mounts at some point, but it should be designed in such a way that we shouldn't need to care.

noahdesu commented:
I suppose the point of convergence is that PVs can be provisioned by a cloud provider automatically (this design) vs provisioned by an administrator manually (current)

jarrpa commented:
Actually, this design can also take into account static provisioning of PVs. If you create a StorageClass with `no-provisioner` and then annotate the static PVs to be part of the StorageClass, you can provision PVCs agaisnt that StorageClass and bind to the static PVs. You don't have a guarantee of which specific PVs bind to which PVC, but that's not something to be concerned about as long as the information can be easily verified after the fact. With `volumeBindingMode: WaitForFirstConsumer` on the StorageClass, the Pod scheduling will also be affected by the availability of a static PV capable of meeting the scheduling requirements of the Pods.

jarrpa commented:
The meeting was held. I think it went well, if a bit long, as good design discussions tend to do. :) The recording is available here for any who are interested: https://www.youtube.com/watch?v=V0mKiFFw-FQ

We came away with a few TODOs from the meeting, the results of which will be communicated in this PR.

leseb commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,69 @@
+struct StorageClassDeviceSet {
+    Name                 string // A unique identifier for the set
+    Count                int // Number of devices in this set
+    Resources            v1.ResourceRequirements // Requests/limits for the devices
```
Does this mean QoS?

jarrpa replied:
It means CPU and memory requests/limits. Kubernetes will use this to 1.) determine if a Pod can be scheduled on the node based on available CPU and memory on the node, and 2.) determine priority when a node starts running out of resources about which pods to start evicting.

leseb commented on design/storage-class-device-set.md:
````diff
@@ -0,0 +1,69 @@
+}
+```
+
+A provider will be able to use the StorageClassDeviceSet struct to describe the properties of a particular set of StorageClassDevices. In this design, the notion of a "StorageClassDevice" is an abstract concept, separate from underlying storage devices. In particular, a single StorageClassDevice could be comprised of multiple underlying storage devices, specified by having more than one item in t ...
````
```suggestion
A provider will be able to use the StorageClassDeviceSet struct to describe the properties of a particular set of `StorageClassDevices`. In this design, the notion of a "StorageClassDevice" is an abstract concept, separate from underlying storage devices. In particular, a single `StorageClassDevice` could be comprised of multiple underlying storage devices, specified by having more than one item in the `volumeClaimTemplates` field.
```

leseb commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,69 @@
+
+A provider will be able to use the StorageClassDeviceSet struct to describe the properties of a particular set of StorageClassDevices. In this design, the notion of a "StorageClassDevice" is an abstract concept, separate from underlying storage devices. In particular, a single StorageClassDevice could be comprised of multiple underlying storage devices, specified by having more than one item in t ...
+
+A StorageClassDeviceSet will have the following fields associated with it:
```
```suggestion
A `StorageClassDeviceSet` will have the following fields associated with it:
```

leseb commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,69 @@
+
+* **name**: A name for the set. **[required]**
+* **count**: The number of devices in the set. **[required]**
+* **resources**: The CPU and RAM requests/limits for the devices. Default is no resource requests.
```
I don't understand the concept of resources for a device. I understand for a pod, but a device? Can someone clarify that to me? Thanks!

noahdesu replied:
The way I understand this from the conversation is that in order to bind a PV to a node may require certain resources. An example of this may be that a very large device may support multiple OSDs, but that implies that the node would need to have the resources capable of supporting such OSDs. This occupied a lot of the discussion, and I believe there is a TODO item to enumerate a covering set of concrete examples that would address your question.

jarrpa replied:
Correct. I will be updating the doc with a (hopefully) better explanation and some examples.

leseb replied:
Ok, makes sense, thanks!

leseb commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,69 @@
+  *  **storageClassName**: The StorageClass to provision PVCs from.
+
+## Example Workflow: rook-ceph OSDs
+The CephCluster CRD could be extended to include a new field: `spec.storage.StorageClassDeviceSets`, which would be a list of one or more StorageClassDeviceSets. If elements exist in this list, the CephCluster controller would then create enough PVCs to match each StorageClassDeviceSet's `Count` field, attach them to individual OsdPrepare Jobs, then attach them to OSD Pods once the Jobs are compl ...
```
```suggestion
The CephCluster CRD could be extended to include a new field: `spec.storage.StorageClassDeviceSets`, which would be a list of one or more `StorageClassDeviceSets`. If elements exist in this list, the `CephCluster` controller would then create enough PVCs to match each StorageClassDeviceSet's `Count` field, attach them to individual `OsdPrepare` Jobs, then attach them to OSD Pods once the Jobs are completed. For the initial implementation, only one `volumeClaimTemplate` would be supporter, if only to tighten the scope for an MVP.
```

leseb commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,69 @@
+If the admin wishes to control device placement, it will be up to them to make sure the desired nodes are labeled properly to ensure the Kubernetes scheduler will distribute the OSD Pods based on Placement criteria.
+
+### OSDs as StatefulSets
+A set of OSDs could be represented by a StatefulSet. This would allow rook-ceph to offload the entire notion of Pod and PVC creation and management to Kubernetes while still maintaining OSD identity.
```
```suggestion
A set of OSDs could be represented by a StatefulSet. This would allow Rook-Ceph to offload the entire notion of Pod and PVC creation and management to Kubernetes while still maintaining OSD identity.
```

leseb commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,69 @@
+### OSDs as StatefulSets
+A set of OSDs could be represented by a StatefulSet. This would allow rook-ceph to offload the entire notion of Pod and PVC creation and management to Kubernetes while still maintaining OSD identity.
+
+Though currently the OSD ID is reflected in the OSD Pod name, this would change because the OSD Pod name would be prescribed to be `<statefulset_name>-<n>`, where `n` is a number from 0 to `Count`. To mitigate this, the OSD ID could be surfaced in the CephCluster CR's `status`, which would still be reflected in a `kubectl get cephcluster` and could be more exclusively fetched using `-o jsonpath`.
```
Looks acceptable to me since we don't need to locate a drive anymore on a specific machine.

noahdesu commented:
The more I think about the inclusion of resource requirements in `StorageClassDeviceSet` the more that I see how it intends to parallel existing real-world abstractions. Here is an example:

Assume you take control of an existing cluster that has nodes and PVs already attached. Further assume that the cluster has been very well spec'd out--disks, cpus, memory, etc... are all well balanced and deliberately chosen. If a _node_ fails in this environment and its disks can be re-purposed into replacement nodes then how would an administrator handle this?

From one point of view we can think about the disk as _requiring something of_ its new future host--that is, the disk _implies_ certain resource requirements on its host. In this light the abstraction makes sense: we want to pair the disk with something _like_ its previous host. If the PV hosted 10 OSDs, we may be able to say something meaningful about the requirements of the host.

But I think ultimately it seems more limiting than useful. Some of the scenarios that I've managed to concoct (which may not be realistic!) are:

* What if a device may host multiple OSDs, but there is some scenario in the future in which a subset of the OSDs should be started on a host? In this case, how would the resource requirements be interpreted?

* If the resource requirements are an _absolute minimum_ then perhaps it would be useful for scheduling, but otherwise I can't formulate a useful definition. Here is an example:

Say a some PV is mounted to a _monster_ machine capable of hosting many OSDs. Are the associated resource requirements for the PV _related at all_ to the host's resource capabilities? If yes, then that seems odd, because I could certainly move the PV to a less capable OSD and still be fine. Which begs the question: if the resource requirements of the PV are independent of a particular host, then what exactly do they represent?

sebastian-philipp commented:
# Drive Groups

Creating real-world OSDs is complex:

* Some configurations deploy multiple OSDs on a single drive
* Some configurations are using more than one drive for a single OSD.
* Deployments often look similar on multiple hosts. 
* There are some advanced configurations possible, like encrypted drives.

All of these setups are valid real-world configurations that need to be supported.

The Ceph project defines a data structure that allows defining groups of drives to be provisioned in a specific way by ceph-volume: Drive Groups.

Drive Groups were originally designed to be ephemeral, but it turns out that orchestrators like DeepSea store them permanently in order to have a source of truth when (re-)provisioning OSDs.

Also, Drive Groups were originally designed to be host specific. But the concept of hosts is not really required for the Drive Group data structure make sense, as they only select a subset of a set of aviailble drives. 

DeepSea has a documentation of [some example drive groups](https://github.com/SUSE/DeepSea/wiki/Drive-Groups#example-drive-group-files).

A complete specificatino is documented in the [ceph documentation](http://docs.ceph.com/docs/master/mgr/orchestrator_modules/#orchestrator.DriveGroupSpec).


## DeviceSet vs DriveGroup

A DeviceSet to provision 8 OSDs on 8 drives could look like:

```yaml
name: "my_device_set"
count: 8
```

The Drive Group would look like so:

```yaml
host_pattern: "hostname1"
data_devices:
  count: 8
```

A Drive Group with 8 OSDs using a shared fast drive could look similar to this:

```yaml
host_pattern: "hostname1"
data_devices:
  count: 8
  model: MC-55-44-XZ
db_devices:
  model: NVME
db_slots: 8
```
## ResourceRequirements and Placement

Drive Groups don't yet provide orchestrator specific extensions, like resource requirements or placement specs, but that could be added trivially.

Also a name could be added to Drive Groups. 

## Are Drive Groups applicable here?

From a data structure perspective, DeviceSets very much resemble Drive Groups.

Using DeviceSets only for AWS deployments seems strange to me as there is no inherent problem with applying DeviceSets also for normal deployments. Especially as other Ceph orchestrators are doing this already: `StorageClassDeviceSets` is very similar to DeepSea's cluster wide static configuration in `/srv/salt/ceph/configuration/files/drive_groups.yml`.

There is just one fundamental thing:

**What is going to be the source of truth for Rook: The reality of OSDs (see Blaine's proposal) or DeviceSets?**

I think both approaches are sensible.

jarrpa commented:
I've updated the design doc significantly. The following are the major points:

* Added the formatting suggestions from @leseb 
* Updated the struct description to include information about how the resources and placements are relevant in the struct
* Added the `DriveGroups` text proposed by @sebastian-philipp, with some minor editing.
* Added some examples to hopefully cover enough use cases to allow people to understand the design reasoning
* Big One: I've removed the idea for using StatefulSets. I had a good talk with Travis to run the mental exercise of considering some of the more unusual edge cases, and decided that it was not worth the hassle of jumping through the technical hoops required to make StatefulSets as robust as just using Deployments.

  The biggest (and simplest) use case that would be difficult to accommodate is multiple OSDs on a single device. To make it work in a StatefulSet, you would need to specify one StatefulSet per device, and then hardcode a single PVC in the Pod spec template for all Pods in the StatefulSet. This completely goes against the intended use case of the StatefulSet. This and some over use cases convinced me to simply deal with the technical overhead of managing PVC lifecycles and stick to using Deployments.

sebastian-philipp commented:
(one minor thing, if OSDs are moved to a different host, who is going to adjust the crushmap?)

jarrpa commented:
> (one minor thing, if OSDs are moved to a different host, who is going to adjust the crushmap?)

One idea is to NOT specify the node name in the CRUSH map, at least for cloud environments, since this could change at any time. A granularity of region and zone is often sufficient, but the admin could always add additional node labels to be picked up by Rook and set in the Deployment configuration. Alternatively, Pod specs to allow for inserting the node and pod name as environment variables to the container, so we could define the CRUSH map to use those variables somehow.

BlaineEXE commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,438 @@
+  ...
+  storage:
+    ...
+    deviceSets:
```
Are we not renaming these to `storageClassDeviceSets`?

jarrpa replied:
Good catch.

BlaineEXE commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,438 @@
+    deviceSets:
+    - name: cluster1-set1
+      count: 3
+      volumeClaimTemplates:
```
None of these examples use more than a single volume claim template, and I'm struggling to imagine how multiple templates might work under the hood. Could you explain how multiple templates would be handled?

jarrpa replied:
In the revamped design here, you would have one Deployment and one Pod per OSD and you just configure the Pod to attach to two PVCs. Then you would have to specify some additional configuration on how you want the devices to be consumed (e.g. one for metadata and one for data).

jarrpa commented:
Updated the design doc to reflect two changes per @BlaineEXE's suggestions:

* Fixed a few typos
* Added a fifth example to show how you could use multiple entries in the `volumeClaimTemplates` field. As with the other ones, this is a Rook-Ceph-specific example. Other providers might be able to come up with different use cases.

BlaineEXE commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,487 @@
+
+* **name**: A name for the set. **[required]**
+* **count**: The number of devices in the set. **[required]**
+* **resources**: The CPU and RAM requests/limits for the devices. Default is no
```
I may have missed where precedence is stated in the doc. Is it stated the override precedence order of the device sets compared to other parts of the PR?

jarrpa commented:
@BlaineEXE Added the following paragraph:
```
In keeping with current Rook-Ceph patterns, the **resources** and **placement**
for the OSDs specified in the `StorageClassDeviceSet` would override any
cluster-wide configurations for OSDs. Additionally, the `StorageClassDeviceSet`
configuration would take precedence over any other conflicting configurations,
such as if `useAllDevices` is configured for the cluster.
```

jarrpa commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,493 @@
+      count: 3
+      config:
+        metadataDevice: "/dev/rook/device1"
+      volumeClaimTemplates:
```
Even the notion of calling something a `dataTemplate` feels like it's getting too tied down to how Rook-Ceph intends to use it. The list of templates is much more generic.

leseb replied:
Having `name` key off sounds good so let's stick with @jarrpa current proposal.

ashishranjan738 commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,493 @@
+    Resources            v1.ResourceRequirements // Requests/limits for the devices
+    Placement            rook.Placement // Placement constraints for the devices
+    Config               map[string]string // Provider-specific device configuration
+    volumeClaimTemplates []v1.PersistentVolumeClaimSpec // List of PVC templates for the underlying storage devices
```
whats the need of a list of PVCSpec template if we are having a count field in the StorageClassDeviceSet

jarrpa replied:
The count would apply to every template in the list, so if you have two templates and a count of 3 you would end up with 6 PVCs. This is intended to accommodate the scenario where one OSD is using multiple devices, and is shown in the last example in the doc.

jarrpa commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,493 @@
+* **volumeClaimTemplates**: A list of PVC templates to use for provisioning the
+  underlying storage devices.
+
+An entry in `volumeClaimTemplates` must specify the following fields:
```
This would vary depending on the operator. For Rook-Ceph we could prescribe the function of each template in the list, with the first template being the data device and the second, if specified, would be the metadata device.

jarrpa commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,493 @@
+## Example Workflow: rook-ceph OSDs
+
+The CephCluster CRD could be extended to include a new field:
+`spec.storage.StorageClassDeviceSets`, which would be a list of one or more
```
I'd like to keep it as-is.

jarrpa commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,493 @@
+controller would then create enough PVCs to match each `StorageClassDeviceSet`'s
+`Count` field, attach them to individual OsdPrepare Jobs, then attach them to
+OSD Pods once the Jobs are completed. For the initial implementation, only one
+entry in `volumeClaimTemplates` would be supported, if only to tighten the scope
```
See my earlier comment about multiple templates in a single device set.

jarrpa commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,493 @@
+1. Create any missing OSD PVCs
+1. Get all matching OSD Deployments
+1. Check that all OSD Deployments are using valid OSD PVCs
+    * If not, probably remove the OSD Deployment?
```
We can avoid this scenario for now, yeah.

jarrpa commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,493 @@
+configuration would take precedence over any other conflicting configurations,
+such as if `useAllDevices` is configured for the cluster.
+
+### OSD Deployment Behavior
```
That would only happen if the claim template is somehow explicitly configured to `volumeNode: Filesystem`. Even then, we could always just ignore the field.

jarrpa commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,493 @@
+  storage devices.
+  * **storageClassName**: The StorageClass to provision PVCs from.
+
+## Example Workflow: rook-ceph OSDs
```
I would say that's beyond the scope of this initial design.

jarrpa commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,493 @@
+recommended directions for [local
+PVs](https://kubernetes.io/blog/2019/04/04/kubernetes-1.14-local-persistent-volumes-ga/).
+
+### Possible Implementation: DriveGroups
```
Will do.

jarrpa commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,493 @@
+
+It should be noted, in the case where the only nodes that can run a new OSD Pod
+is are nodes with OSD Pods already on them, one of those nodes would be
+selected. In addition, EBS volumes may not cross between AZs once greated, so a
```
Yup.

jarrpa commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,493 @@
+    storageClassDeviceSet:
+    - name: cluster1-set1
+      count: 3
+      volumeClaimTemplates:
```
The template starts on the next line with `- resources:`.

jarrpa commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,493 @@
+  storage:
+    ...
+    nodes:
+    - name: node1
```
Does Rook-Ceph not to anything to label/identify nodes that are explicitly specified in the CephCluster CR?

jarrpa replied:
I guess I can live without it.

jarrpa commented:
Updated the doc again to fix a typo and clarify the OSD examples.

leseb commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,478 @@
+1. Get all matching OSD PVCs
+1. Create any missing OSD PVCs
+1. Get all matching OSD Deployments
+1. Check that all OSD Deployments are using valid OSD PVCs
```
It is not clear to me, what will an invalid PVCs be?

jarrpa replied:
Any PVC that doesn't match the template specified in the device set.

leseb commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,478 @@
+    * Remove any PVCs used by OSD Deployments from the list of PVCs to be
+    worked on
+1. Run an OsdPrepare Job on all unused and uninitialized PVCs
+    * This may be either one Job per PVC or one Job per batch of PVCs. One per
```
Yes, if we batch and one prepare fails then the remaining ones won't be prepared so 1 prepare per PVC.

jarrpa replied:
Sure,

leseb commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,493 @@
+* Some configurations deploy multiple OSDs on a single drive
+* Some configurations are using more than one drive for a single OSD.
+* Deployments often look similar on multiple hosts.
+* There are some advanced configurations possible, like encrypted drives.
```
Every permutation results in a different argument passed to ceph-volume ;-)

leseb commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,478 @@
+know if a given node can support running an OSD process.
+
+It should be noted, in the case where the only nodes that can run a new OSD Pod
+is are nodes with OSD Pods already on them, one of those nodes would be
```
```suggestion
are nodes with OSD Pods already on them, one of those nodes would be
```

leseb commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,478 @@
+Given the complexity of this design, here are a few examples to showcase
+possible configurations for OSD `StorageClassDeviceSets`.
+
+#### Example 1: AWS cross-AZ
```
Given Ceph's synchronous's nature, I wouldn't put PVC across AZ as the latency will increase and overall performance will be badly impacted.
This could make sense if we were able to be more flexible in the way we configure the CRUSH map and then create pool per AZ and not span across AZ.
So I'd be in favor of not allowing that for the initial MVP.

jarrpa replied:
It would be more work to disallow it than allow it, and for an MVP I would rather just be lax. Allow it and let the admin suffer any consequences.

leseb replied:
In that case, we should have removed it and/or document this is **not** a valid deployment scenario...

jarrpa commented:
Updated the design with the following:

* Fixed a typo
* Changed the `volumeClaimTemplates` field to be a list of `PersistentVolumeClaim` instead of `PersistentVolumeClainSpec`. @ashishranjan738 Correctly pointed out that we need to uniquely track which template a PVC belongs to, and he pointed out that StatefulSets solve this by using the `metadata.name` field as a means to generate a unique name for the PVC. We can also always label the PVC with the set and template names.

jarrpa commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,487 @@
+An entry in `volumeClaimTemplates` must specify the following fields:
+  * **resources.requests.storage**: The desired capacity for the underlying
+  storage devices.
+  * **storageClassName**: The StorageClass to provision PVCs from.
```
Sure, adding.

jarrpa commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,487 @@
+
+The CephCluster CRD could be extended to include a new field:
+`spec.storage.StorageClassDeviceSets`, which would be a list of one or more
+`StorageClassDeviceSets`. If elements exist in this list, the CephCluster
```
"one or more `StorageClassDeviceSet`" is grammatically awkward, and "`StorageClassDeviceSet`s" just looks weird.

jarrpa replied:
I think it'll be fine. I mean, heck, the title of this doc is `# Rook StorageClassDeviceSets`. ;)

jarrpa commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,487 @@
+for the OSDs specified in the `StorageClassDeviceSet` would override any
+cluster-wide configurations for OSDs. Additionally, the `StorageClassDeviceSet`
+configuration would take precedence over any other conflicting configurations,
+such as if `useAllDevices` is configured for the cluster.
```
All right.

jarrpa commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,487 @@
+        spec:
+          resources:
+            requests:
+              storage: 1
```
Nope. 1 is technically just one byte. Alternatively we should just omitted and, if not specified, use a value of `1` on the backend.

jarrpa commented on design/storage-class-device-set.md:
```diff
@@ -0,0 +1,487 @@
+entire storage device and get that capacity regardless of what is set for the
+PVC.
+
+#### Example 5: Multiple devices per OSD
```
All right!

jarrpa commented:
@travisn Updated the design based on your recent comments.
